bee-signing/src/ternary/mss.rs

```diff
@@ -0,0 +1,407 @@
+                .copy_from(&self.tree[sibling_index * HASH_LENGTH..(sibling_index + 1) * HASH_LENGTH]);
+            i += 1;
+        }
+
+        self.index += 1;
+
+        Ok(Self::Signature::from_trits(state)
+            .map_err(|_| Error::SignatureGenerationFailed)?
+            .index(self.index - 1))
+    }
+}
+
+/// Merkle Signature Scheme public key.
+pub struct MssPublicKey<S, K> {
+    state: TritBuf<T1B1Buf>,
+    depth: Option<u8>,
+    sponge: PhantomData<S>,
+    key: PhantomData<K>,
+}
+
+impl<S, K> MssPublicKey<S, K>
+where
+    S: Sponge + Default,
+    K: PublicKey,
+{
+    /// Sets the depth of the public key.
+    pub fn depth(mut self, depth: u8) -> Self {
+        self.depth.replace(depth);
+        self
+    }
+}
+
+impl<S, K> PublicKey for MssPublicKey<S, K>
+where
+    S: Sponge + Default,
+    K: PublicKey,
+    <K as PublicKey>::Signature: RecoverableSignature,
+{
+    type Signature = MssSignature<S>;
+    type Error = Error;
+
+    fn verify(&self, message: &Trits<T1B1>, signature: &Self::Signature) -> Result<bool, Self::Error> {
+        if signature.size() % SIGNATURE_FRAGMENT_LENGTH != 0 || signature.size() < 2 * SIGNATURE_FRAGMENT_LENGTH {
+            return Err(Error::InvalidSignatureSize);
+        }
+
+        let depth = self.depth.ok_or(Error::MissingDepth)?;
+        let signature_index = signature.index.ok_or(Error::MissingSignatureIndex)?;
+
+        let mut sponge = S::default();
+        let underlying_signature = K::Signature::from_trits(
+            signature.state[0..((signature.state.len() / SIGNATURE_FRAGMENT_LENGTH) - 1) * SIGNATURE_FRAGMENT_LENGTH]
+                .to_buf(),
+        )
+        .map_err(|_| Error::FailedUnderlyingSignatureGeneration)?;
+
+        // Safe to unwrap since we already validated the signature size.
+        let siblings: &Trits<T1B1> = signature.state.chunks(SIGNATURE_FRAGMENT_LENGTH).last().unwrap();
+        // Recovers underlying public key from underlying signature.
+        let underlying_public_key = underlying_signature
+            .recover_public_key(message)
+            .map_err(|_| Self::Error::FailedUnderlyingPublicKeyRecovery)?;
+        let mut hash = TritBuf::<T1B1Buf>::zeros(HASH_LENGTH);
+
+        hash.copy_from(underlying_public_key.as_trits());
+
+        // Hash the underlying public key with the merkle branch to recover the merkle root.
+        let mut j = 1;
+        for (i, sibling) in siblings.chunks(HASH_LENGTH).enumerate() {
+            if depth - 1 == i as u8 {
+                break;
+            }
+
+            if signature_index & j != 0 {
+                sponge.absorb(sibling).map_err(|_| Self::Error::FailedSpongeOperation)?;
+                sponge.absorb(&hash).map_err(|_| Self::Error::FailedSpongeOperation)?;
+            } else {
+                sponge.absorb(&hash).map_err(|_| Self::Error::FailedSpongeOperation)?;
+                sponge
+                    .absorb(&sibling)
```

Comment:
Why is `sibling` taken by reference here and not above? It also seems to be the cause of the bizarre formatting.

Reply:
Good catch, `sibling` is already a reference. Removing it.